### ISSUE TITLE
Feature request: Variable access to the client cert in PEM format

### ISSUE THREAD

**Issue by fiddyspence** [haproxy/haproxy#20]
Hi - loving your work.

It would be useful for my application (putting HAProxy in front of a service that wants the PEM formatted certificate as an HTTP header, in my case 'X-Client-Cert') for an additional variable, like https://cbonte.github.io/haproxy-dconv/configuration-1.5.html#7.3.4-ssl_c_der, but the client certificate in PEM format.  Pretty please.

**Comment by figless** [comment]
What about:
"set-header X-Client-Cert -----BEGIN\ CERTIFICATE-----\ %[ssl_c_der,base64]\ -----END\ CERTIFICATE-----\ #"
Referenced from http://comments.gmane.org/gmane.comp.web.haproxy/18796 with haproxy 1.5.7+

**Comment by MattPark** [comment]
Yeah you can probably close this one, it *mostly* works.
(slight changes from spaces to newlines)
`set-header X-Client-Cert \n-----BEGIN\ CERTIFICATE-----\n%[ssl_c_der,base64]\n-----END\ CERTIFICATE-----\n`
The part that doesn't work is that base64 encoded PEMs need line breaks every 64 chars.  This solution puts the whole base64 piece on one line, which openssl doesn't like.

**Comment by wtarreau** [comment]
On Tue, Jun 14, 2016 at 08:40:53PM -0700, Matt Park wrote:
> Yeah you can probably close this one, it *mostly* works.
> (slight changes from spaces to newlines)
> `set-header X-Client-Cert \n-----BEGIN\ CERTIFICATE-----\n%[ssl_c_der,base64]\n-----END\ CERTIFICATE-----\n`
> The part that doesn't work is that base64 encoded PEMs need line breaks every
> 64 chars.  This solution puts the whole base64 piece on one line, which
> openssl doesn't like.

Changing spaces to newlines is pointless as you're sending this in
headers and HTTP explicitly states that 1) multi-line headers should
not be sent, and 2) that any intermediary must fix them by replacing
line breaks with spaces.

Your server *needs* to convert the cert correctly. One thing that could
possibly be done would be to encode the whole block using base64, but
normally what people do is to use a different character for line breaks
(eg: '|').

Also, since you'll have to cut the base64 into pieces by yourself on the
server side, I really think that it's much safer to send only the
base64-encoded cert in the header and to convert it to PEM on the server
before passing it to openssl. There you can prepend the BEGIN CERTIFICATE
header, cut the base64 into 64 chars lines, append the END CERT and pass
it to openssl. That's much safer, simpler and even more portable.

**Comment by MattPark** [comment]
Yeah I mean the RFC says ["Header fields can be extended over multiple lines"](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4) which in my mind is pretty clear and explicit so there must be some conflicting guidance.  

I don't know what requirement @fiddyspence has.  Not everyone has the capability to change the application to meet the headers, esp in the case of legacy applications where you are given a spec and have to adjust everything in between to meet the spec.

I usually handle authentication at the proxy and just pass back a user identifier, but again if your app cannot be modified to accept that it may not help.

**Comment by wtarreau** [comment]
On Thu, Jun 16, 2016 at 09:32:25AM -0700, Matt Park wrote:
> Yeah I mean the RFC says ["Header fields can be extended over multiple
> lines"](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4) which in
> my mind is pretty clear and explicit so there must be some conflicting
> guidance.  

2616 is 17 years old and obsoleted by 7230. the 2616 makes it very clear that 
multi-line may be folded into a single line by anyone along the chain.

> I don't know what requirement @fiddyspence has.  Not everyone has the
> capability to change the application to meet the headers, esp in the case of
> legacy applications where you are given a spec and have to adjust everything
> in between to meet the spec.

No but if the application relies on random behaviour like "let's hope that
a multiline header will arrive unfolded", it must be called bogus and there
is a moment where a bug is going to strike. You can work-around bugs, but the
more you workaround them, the higher they cost. If in your case it works, it's
fine, but keep in mind that this might be the very last warning from a product
and that in a few years you won't be able at all to rely on this behaviour.
